### PR TITLE
Pro: History

### PR DESCRIPTION
**Pull request by rhymes**
## What type of PR is this? (check all applicable)

- [ ] Refactor
- [x] Feature
- [ ] Bug Fix
- [ ] Documentation Update

## Description

With this PR we introduce a new feature only available to pro users: the history.

The feature borrows a lot from the reading list, maybe in the future we could refactor both to share more code (both in the CSS which is nearly identical, the JS and the Ruby code). I don't think this PR should do that right away.

### Index

As for the reading list we're using Algolia for indexing. The index is built by instances of the `PageView` (only those of pro users are added to the index):

This is the index's definition:

```ruby
  algoliasearch index_name: "UserHistory", per_environment: true, if: :belongs_to_pro_user? do
    attributes :referrer, :time_tracked_in_seconds, :user_agent, :article_tags

    attribute(:article_title) { article.title }
    attribute(:article_path) { article.path }
    attribute(:article_reading_time) { article.reading_time }
    attribute(:viewable_by) { user_id }
    attribute(:visited_at_timestamp) { created_at.to_i }

    attribute :article_user do
      # ...
    end

    attribute :readable_visited_at do
      # ...
    end

    searchableAttributes(
      %i[referrer user_agent article_title article_searchable_tags article_searchable_text],
    )

    tags { article_tags }

    attributesForFaceting ["filterOnly(viewable_by)"]

    customRanking ["desc(visited_at_timestamp)"]
  end
```

Basically we index the referrer (currently unused), the time tracked (currently unused), the user agen (currently unused), some properties for the article (tags, title, path, reading time, user). We also index a readable date shown (which is not the time the article was published like in the reading list but the time the user visited such article).

We also allow the user to search  on article's tags, title and body excerpt. The searchable index currently contains the referrer and the user agent as well BUT I'm not sure those should be "silently" searchable, without a better UI (or a visible facet?). 

Everything is then ranked by descending timestamp, so that the new visits will appear at the top

### Pagination

This PR contains pagination for the results, see screenshots below

#### Open questions / issues

* should we add referrer, time tracked and  user agent if we don't show them? If we do want to show them, what's the best way to do it?

* should the user be able to search on referrer and user agent "silently" ?

* what else we should add to the single item in the UI?

* Page views can be repeated, in the sense that if I visit an article twice in a row, I'll see two entries. Is that ok? See example:

Screenshot 2019-06-18 at 1 15 25 PM

* should we put the link to the history page in the homepage?

### How it looks

Screenshot_2019-06-18 History - The DEV(local) Community

### Pagination and search demo

history

## Added to documentation?

- [ ] docs.dev.to
- [ ] readme
- [x] no documentation needed

### CODE REVIEW

File: app/assets/stylesheets/history.scss [hunk 1]
```diff
@@ -0,0 +1,237 @@
+@import 'variables';
```
Comment (rhymes) [inline app/assets/stylesheets/history.scss]: This file is essentially `readinglist.css` with the names of the classes replaced and one additional class for the pagination button.

File: app/assets/stylesheets/scaffolds.scss [hunk 2]
```diff
@@ -57,6 +59,7 @@ body {
   &.stories-search,
   &.podcast_episodes-index,
   &.reading_list_items-index,
+  &.history-index,
```
Comment (rhymes) [inline app/assets/stylesheets/scaffolds.scss]: added history index to here, we should probably use a common class for all of these containers, instead of listing them explicitly, it took me a while to understand why there was no margin :D

File: app/javascript/history/history.jsx [hunk 3]
```diff
@@ -0,0 +1,212 @@
+import { h, Component } from 'preact';
+import { PropTypes } from 'preact-compat';
+import debounce from 'lodash.debounce';
```
Comment (rhymes) [inline app/javascript/history/history.jsx]: I'm using `lodash.debounce` to avoid sending every keystroke to Algolia. It's configured to debounce up to 300ms

File: app/javascript/history/history.jsx [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+      index: null,
+      itemsLoaded: false,
+      hitsPerPage: 100,
+      availableTags,
+      selectedTags: [],
+      page: 0,
+      showNextPageButton: false,
+    };
+
+    this.handleTyping = debounce(this.handleTyping.bind(this), 300, {
```
Comment (rhymes) [inline app/javascript/history/history.jsx]: where the debouncing magic happens

File: app/javascript/history/history.jsx [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+      leading: true,
+    });
+  }
+
+  componentDidMount() {
+    const index = setupAlgoliaIndex({
```
Comment (rhymes) [inline app/javascript/history/history.jsx]: I plan to use this utility function in the future, because I've noticed the algolia init code is repeated everywhere

File: app/javascript/packs/history.jsx [hunk 6]
```diff
@@ -0,0 +1,22 @@
+import { h, render } from 'preact';
+import { getUserDataAndCsrfToken } from '../chat/util';
+import { History } from '../history/history';
+
+function loadComponent() {
+  getUserDataAndCsrfToken().then(({ currentUser }) => {
```
Comment (rhymes) [inline app/javascript/packs/history.jsx]: I think it makes more sense to get the user data before, instead of using a `setInterval` function like in the reading list. This way the component just accepts the availableTags from the outside and that's it

File: app/views/articles/_sidebar_nav.html.erb [hunk 7]
```diff
@@ -19,17 +19,17 @@
         navigation
       </div>
       <div class="sidebar-nav-element">
-        <a class="sidebar-nav-link" href="/readinglist?we-hope-you-like-the-new-and-improved-reading-list---more-positive-changes-coming=woohoo">
+        <a class="sidebar-nav-link" href="<%= readinglist_path %>">
```
Comment (rhymes) [inline app/views/articles/_sidebar_nav.html.erb]: This refactoring is because I was going to add the href to the history page in the homepage but decided against it. Not sure if we should put the link for pro users here.

File: config/routes.rb [hunk 8]
```diff
@@ -44,7 +44,7 @@
         post "save_status"
       end
     end
-    resources :tags, only: %i[index show edit update]
+    resources :tags, only: %i[index edit update]
```
Comment (rhymes) [inline config/routes.rb]: show doesn't exist in that controller

File: package.json [hunk 9]
```diff
@@ -21,8 +21,8 @@
   },
   "lint-staged": {
     "*.{js,jsx}": [
-      "eslint --fix",
       "prettier --write",
+      "eslint --fix",
```
Comment (rhymes) [inline package.json]: inverted these two, right now you still get a lot of non fixed code because of the order

**Comment by rhymes** [comment]
@benhalpern thanks for releasing it right away :) I forgot to add a script to re-index all pro user's pageviews. Can you do it manually?

**Comment by benhalpern** [comment]
should we add referrer, time tracked and user agent if we don't show them? If we do want to show them, what's the best way to do it?

**Referrer is probably not necessary. I could see future use-cases where that could be neat to help contextually recall an old post. But I wouldn't overly concern with that.**

should the user be able to search on referrer and user agent "silently" ?

**I wouldn't worry about it just yet.**

what else we should add to the single item in the UI?

Page views can be repeated, in the sense that if I visit an article twice in a row, I'll see two entries. Is that ok? See example:

**It probably makes sense to not display back-to-back page views that are the same. As a general rule. If we display time spent, maybe they should be summed. I feel like this all would be best handled on the client. Check to see if the item is the same as the previous one and tack it on if that's the case, otherwise move on.**

I'm very excited about this and I think we'll figure out more with use. And since it's not for general use, we'll be able to reindex more often without it being a major engineering ordeal.

**Comment by rhymes** [comment]
@benhalpern I did some experiments

> It probably makes sense to not display back-to-back page views that are the same. As a general rule. If we display time spent, maybe they should be summed. I feel like this all would be best handled on the client. Check to see if the item is the same as the previous one and tack it on if that's the case, otherwise move on.

I'm not sure we should use that metric at all for now. Time spent is measured only occasionally using a random function as a flag, which means that most page views are never updated:

https://github.com/thepracticaldev/dev.to/blob/c413cf823518b24b589d79ca12bd6195cd77587d/app/assets/javascripts/initializers/initializeBaseTracking.js.erb#L118-L122

We can revisit all of this logic (how page views are tracked, stored and updated) in the future.

I've also tried to use client side de-duplication but there are two main problems with it: 

* pages can contain a different amount of items once de-duplicated obviously
* the total count becomes incorrect and computing it at runtime accross different pages becomes hard, because, for example, on the first page I know that out of 100 items 10 are duplicated, so I display 90, but then on the second page i find another duplicated item, so I have to display 89? It would also be super weird for the user to see the total count change

I'm leaving down here a demo of how potentially odd pagination would look like (I've changed the page size to 10 to show the issue):

deduplication

The duplicate visits you see are shown because they occurr on the boundary of different pages.

I found out Algolia has a way to deduplicate results through [grouping](https://www.algolia.com/doc/guides/managing-results/refine-results/grouping/). 

This is what happens if you set it do `distinct: true` on the article path:

**Before**

Screenshot_2019-06-19 History - The DEV(local) Community

**After I re-visit one of the articles in the list**

Screenshot_2019-06-19 History - The DEV(local) Community(1)

As you can see the article count is now changed, indicating I've visited 6 unique articles in my lifetime, because it's basically a `COUNT(distinct)` on the path.

This is what happens if you set it to `distinct: 2` which changes the logic of whole query by returning `hitsPerPage * 2`, in that case we can't show the items count to the user because it becomes meaningless. Example:

Screenshot_2019-06-19 History - The DEV(local) Community(2)

Basically we have three possibilities:

1. de-duplication on the client side
1. de-duplication on the server side using distinct
1. grouping of duplicate results

Let me know which one you'd prefer, I'm not keen on doing it on the client side and just returning unique visits could be okay, if I visit the same article twice it pops at the top of the list the next time I visit the history, that's it.

> I'm very excited about this and I think we'll figure out more with use. And since it's not for general use, we'll be able to reindex more often without it being a major engineering ordeal.

:-))

**Comment by rhymes** [comment]
BTW #2, the distinct, it's also how the Firefox history works, if you visit again the same URL, it gets moved at the top of the list and it disappears from where it was before.
